client/service-container-chain/src/data_preservers.rs

```diff
@@ -0,0 +1,420 @@
+            let mut set = self.state.lock().unwrap();
+
+            let event = SpawnerEvent::Stopped(container_chain_para_id, keep_db);
+
+            assert!(set.remove(&event), "Unexpected event {event:?}");
+
+            None
+        }
+    }
+
+    #[tokio::test]
+    async fn task_logic_works() {
+        let spawner = MockSpawner::new();
+
+        let profile_id = 0;
+        let para_id1 = ParaId::from(1);
+        let para_id2 = ParaId::from(2);
+
+        tokio::spawn(task_watch_assignment(spawner.clone(), profile_id));
+        // Wait for task to start and subscribe to block stream.
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        spawner.set_expectations(vec![SpawnerEvent::Started(para_id1, false)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::Active(para_id1));
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![SpawnerEvent::Stopped(para_id1, false)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::NotAssigned);
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![SpawnerEvent::Started(para_id2, false)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::Active(para_id2));
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![SpawnerEvent::Stopped(para_id2, true)]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::Inactive(para_id2));
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::Inactive(para_id1));
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
+
+        spawner.set_expectations(vec![]);
+        spawner.chain_interface.mock_block({
+            let mut map = BTreeMap::new();
+            map.insert(profile_id, DataPreserverAssignment::NotAssigned);
+            map
+        });
+        tokio::time::sleep(Duration::from_millis(100)).await;
+        spawner.ensure_all_events_were_emitted();
```

Comment:
Active(1) -> Active(2) case is not tested.

Reply:
fixed